rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,340 @@
+As long as state ownership is isolated, we can do something like this:
+
+```ts
+const embeddableMigrator: PersistableStateMigrationFn (state: { type: string; input: unknown }, version: string): State {
+  return {
+    ...migrateInputToLatest(state.input),
+    enhancements: {
+       ...state.input.enhancements.map(enhancement => 
+          persistableStateMigrations.get(enhancement).migrate(state.input.enhancements[extension], version)),
+    },
+    specializedState: persistableStateMigrations.get(state.type).migrate(state.input.specializedState, version)),
+  }
+}
+```
+
+The key here is that we have to separate out the state owned by VisualizeEmbeddableInput and that owned by EmbeddableInput. If we allow them both access to the full input (VisualizeEmbeddableInput extends EmbeddableInput), there is the possibility for key clashes. For instance:
+
+```ts
+// We start out with the base class having an attribute of `title`. 
+interface EmbeddableInputV1 {
+  title: string;
+}
+
+// `Title` is taken by the base class, so Visualize plugin adds `defaultTitle`.
+interface VisualizeEmbeddableInputV1 implements EmbeddableInputV1 {
+  defaultTitle: string;
+}
+
+// In a future version, EmbeddableInput author decides to rename `title` => `defaultTitle`.
+// They don't know about `VisualizeEmbeddableInput` so don't realize there is going to be
+// a key collision.
+interface EmbeddableInputV2 {
+  defaultTitle: string;
+}
+```
+
+It's probably a rare occurrence, but it's also very risky. EmbeddableInput thinks it owns defaultTitle. Let's say it decides to change the name yet again, so in V3 changes it to customPanelTitle and removes defaultTitle. Let's also say VisualizeEmbeddable author doesn't add a migration for V2 or V3. A user migrates to V3, VisualizeEmbeddable just had its defaultTitle state wiped out unknowingly (types will still say it's a required parameters). This has the potential to permanently break saved objects with no way to undo the migration.
+
+# What saved object migrations 
+
+With saved object migrations (or any other migration of potentially high amount of state objects) currently we try to detect when a migration needs to be performed. For example if there was no change in specific saved object between minors we will not try to read or write those objects from/to elasticsearch.
+
+With extention points to which 3rd party developers might register their own items which expose state that is not possible. For example dashboard saved object might contain state from any embeddable. As we don't know about all those embeddables we need to always fetch all saved objects and run migrations on all of them. 
+If expecting to handle large amount of state objects you should always deep compare input and output state and filter out objects without changes to avoid to many updates in the database.
```

Comment:
Not sure to really follow here. From what I understood, this persistable state service / registry was not directly linked to SO migrations (see `# Server or Client ?` part), and actual migration of the underlying data would only be performed during object access? Why would we need to fetch/migrate everything during SO migration then?

Reply:
if we were to use this for saved object migrations: with saved object migrations we want to know if there is a migration to perform or not. however this is getting harder and harder because of deep dependency trees:

- do we have a migration for dashboard saved objects for 7.9.1 ?:
   - do we have migration for base embeddable input for 7.9.1 ?
      - do we need to migrate filters ?
      - do we need to migrate query ?
      - do we need to migrate timerange ?
   - do we have migration for any of the embeddable types for 7.9.1 ?
       - lens embeddable:
          - do we need to migrate lens state ?
          - do we need to migrate expression ?
              - does any expression function need to be migrated ?
       - visualize embeddable:
          - do we have migration for search source ?
          - do we have migration for index pattern ?
          - do we have migration for any vis type ?
            - once any vis type starts storing expression:
                - is there a migration required for expression?
                  - is there migration required for any expression function ?
    - do we have migration for any enhancement ?
      - drilldowns:
         - do we need to migrate filters ?

you can see how our state is deeply nested with various plugins owning various parts of it. Its also not showing everything (its oversimplified). so figuring out if there is a migration required for specific version (7.9.1) is a hard problem.

also important to note: this is a hard problem with or without global registry. actually without global registry this becomes even harder problem to solve imo, as we don't have registries for a lot of parts of the above dashboard state.